src/Workspaces/Remote/ServiceHub/Services/NavigateToSearch/RemoteNavigateToSearchService.cs

```diff
@@ -47,75 +51,93 @@ public ValueTask HydrateAsync(Checksum solutionChecksum, CancellationToken cance
             return RunServiceAsync(solutionChecksum, solution => ValueTaskFactory.CompletedTask, cancellationToken);
         }
 
-        public ValueTask SearchDocumentAsync(
+        public IAsyncEnumerable<RoslynNavigateToItem> SearchDocumentAsync(
             Checksum solutionChecksum,
             DocumentId documentId,
             string searchPattern,
             ImmutableArray<string> kinds,
-            RemoteServiceCallbackId callbackId,
             CancellationToken cancellationToken)
         {
-            return RunServiceAsync(solutionChecksum, async solution =>
+            return StreamWithSolutionAsync(solutionChecksum, SearchDocumentWorkerAsync, cancellationToken).WithJsonRpcSettings(
+                new JsonRpcEnumerableSettings { MaxReadAhead = MaxReadAhead });
```

Comment:
WithJsonRpcSettings allows us to configure this stream.  This basically allows the server to continuing processing (even in the absence of a client-pull), producing up to MaxReadAhead values while the client is processing the last batch of items returned.  When the client calls back in, it will always grab however many items have been produced so far (or it will wait for any to be produced).